tests/decap/test_subnet_decap.py

```diff
@@ -0,0 +1,226 @@
+        )
+
+    elif ip_version == "IPv6":
+        outer_dst_ipv6 = OUTER_DST_IP_V6
+        if stage == "positive":
+            outer_src_ipv6 = "fc01::10"
+        elif stage == "negative":
+            outer_src_ipv6 = "fc01::10:10"
+
+        inner_packet = testutils.simple_tcpv6_packet(
+            ipv6_src="1::1",
+            ipv6_dst="2::2"
+        )[packet.IPv6]
+        outer_packet = testutils.simple_ipv6ip_packet(
+            eth_dst=eth_dst,
+            eth_src=eth_src,
+            ipv6_src=outer_src_ipv6,
+            ipv6_dst=outer_dst_ipv6,
+            inner_frame=inner_packet
+        )
+
+    return outer_packet
+
+
+def build_expected_vlan_subnet_packet(encapsulated_packet, ip_version, stage, decrease_ttl=False):
+    if stage == "positive":
+        if ip_version == "IPv4":
+            pkt = encapsulated_packet[packet.IP].payload[packet.IP].copy()
+        elif ip_version == "IPv6":
+            pkt = encapsulated_packet[packet.IPv6].payload[packet.IPv6].copy()
+        # Use dummy mac address that will be ignored in mask
+        pkt = packet.Ether(src="aa:bb:cc:dd:ee:ff", dst="aa:bb:cc:dd:ee:ff") / pkt
+    elif stage == "negative":
+        pkt = encapsulated_packet.copy()
+
+    if ip_version == "IPv4":
+        pkt.ttl = pkt.ttl - 1 if decrease_ttl else pkt.ttl
+    elif ip_version == "IPv6":
+        pkt.hlim = pkt.hlim - 1 if decrease_ttl else pkt.hlim
+
+    exp_pkt = Mask(pkt)
+    exp_pkt.set_do_not_care_packet(packet.Ether, "dst")
+    exp_pkt.set_do_not_care_packet(packet.Ether, "src")
+    if ip_version == "IPv4":
+        exp_pkt.set_do_not_care_packet(packet.IP, "chksum")
+    return exp_pkt
+
+
+def verify_packet_with_expected(ptfadapter, stage, pkt, exp_pkt, send_port,
+                                recv_ports=[], recv_port=None, timeout=10, skip_traffic_test=False):    # noqa F811
+    if skip_traffic_test is True:
+        logger.info("Skip traffic test")
+        return
+    ptfadapter.dataplane.flush()
+    testutils.send(ptfadapter, send_port, pkt)
+    if stage == "positive":
+        testutils.verify_packet_any_port(ptfadapter, exp_pkt, recv_ports, timeout=10)
+    elif stage == "negative":
+        testutils.verify_packet(ptfadapter, exp_pkt, recv_port, timeout=10)
+
+
+@pytest.mark.parametrize("ip_version", ["IPv4", "IPv6"])
+@pytest.mark.parametrize("stage", ["positive", "negative"])
+def test_vlan_subnet_decap(rand_selected_dut, tbinfo, ptfhost, ptfadapter, ip_version, stage, skip_traffic_test):  # noqa F811
+    ptf_src_port, downstream_port_ids, upstream_port_ids = prepare_vlan_subnet_test_port(rand_selected_dut, tbinfo)
+
+    encapsulated_packet = build_encapsulated_vlan_subnet_packet(ptfadapter, rand_selected_dut, ip_version, stage)
+    exp_pkt = build_expected_vlan_subnet_packet(encapsulated_packet, ip_version, stage, decrease_ttl=True)
+
+    ptf_target_port = random.choice(downstream_port_ids)
+    setup_arp_responder(ptfhost, ip_version, stage, ptf_target_port)
```

Comment:
better to put the arp responder setup/teardown into a fixture.

Reply:
Fixed